### PR TITLE
Added hierarchical shard selector

### PR DESCRIPTION
**Pull request by r0goyal**

### CODE REVIEW

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 1]
```diff
@@ -277,8 +244,29 @@
                     <value>true</value>
                 </property>
             </activation>
+            <distributionManagement>
```
[the remark this answers is not in the document]
Comment (r0goyal) [reply pom.xml:247]: Yes. These POM changes would be reverted

File: dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        String environment = criteria.getEnvironment();
+
+        if (Objects.equals(environment, all)) {
+            return allNodes(serviceNodes);
+        }
+
+        while (!Strings.isNullOrEmpty(environment)) {
+            val currentEnvNodes = serviceNodes.get(ShardInfo.builder()
+                    .environment(environment)
+                    .build());
```
[the remark this answers is not in the document]
Comment (r0goyal) [reply dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java:37]: Other option is to clone the incoming ShardInfo and then set the environment to the value derived here.

File: dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            if (!currentEnvNodes.isEmpty()) {
+                return currentEnvNodes;
+            }
+
+            if (!environment.contains(separator)){
+                return Lists.newArrayList();
```
[the remark this answers is not in the document]
Comment (r0goyal) [reply dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java:43]: Done

File: pom.xml [hunk 4]
```diff
@@ -208,27 +197,13 @@
         </pluginManagement>
         <plugins>
             <plugin>
-                <groupId>org.codehaus.mojo</groupId>
```
[the remark this answers is not in the document]
Comment (r0goyal) [reply pom.xml:211]: This plugin has not been updated for a while now. Replacing this with jacoco for code coverage

File: dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class HierarchicalEnvironmentAwareShardSelector implements ShardSelector<ShardInfo, MapBasedServiceRegistry<ShardInfo>> {
+
+    private static final String separator = ".";
```
[the remark this answers is not in the document]
Comment (r0goyal) [reply dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java]: Done

File: dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (r0goyal) [reply dropwizard-service-discovery-client/src/main/java/io/appform/dropwizard/discovery/client/selector/HierarchicalEnvironmentAwareShardSelector.java:38]: @santanusinha Would that be a good way to achieve this ?

**Review by santanusinha — APPROVED** [review]

**Pull request merged**